app/models/gem_content_entry.rb

```diff
@@ -0,0 +1,123 @@
+class GemContentEntry
+  SIZE_LIMIT = 500.megabyte
+
+  class << self
+    def entries_from_package(package)
+      GemPackageEnumerator.new(package).filter_map do |tar_entry|
+        Rails.error.handle(context: { gem: package.spec.full_name, entry: tar_entry.full_name }) do
+          from_tar_entry(tar_entry)
+        end
+      end
+    end
+
+    def from_tar_entry(entry)
+      attributes = {
+        size: entry.size,
+        path: entry.full_name,
+        file_mode: entry.header.mode.to_fs(8)
+      }
+
+      if entry.symlink?
+        new(linkname: entry.header.linkname, **attributes)
+      elsif entry.size > SIZE_LIMIT
+        head = entry.read(4096)
+        mime = Magic.buffer(head, Magic::MIME)
+        new(mime: mime, **attributes)
+      else
+        # read immediately because we're parsing a tar.gz and it shares a single IO across all entries.
+        body = entry.read || ""
+        new(
+          body: body,
+          mime: Magic.buffer(body, Magic::MIME),
+          sha256: Digest::SHA256.hexdigest(body),
+          **attributes
+        )
+      end
+    end
+
+    def from_metadata(metadata, &)
+      return unless metadata.respond_to?(:to_h)
```

Comment:
Maybe. I intended to gracefully return nils when there is no metadata, but I think that's handled by the caller and it makes this method too weak. I'll change it.

Reply:
I decided to have this raise InvalidMetadata if it comes in with keys missing. This would indicate corrupted data on S3 and would need to be fixed by reuploading.